WebDAV error: DavException: Received redirect from HTTPS to HTTP

Backup fails with the following error:

```
06-02 11:42:40.630 30094 30665 D BackupCoordinator: Resetting backup metadata...
06-02 11:42:40.631 30094 30665 I BackupCoordinator: Initialize Device!
06-02 11:42:40.645 30094 30665 E BackupCoordinator: Error initializing device
06-02 11:42:40.645 30094 30665 E BackupCoordinator: java.io.IOException: at.bitfire.dav4jvm.exception.DavException: Received redirect from HTTPS to HTTP
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at com.stevesoltys.seedvault.plugins.webdav.WebDavStoragePlugin.initializeDevice(WebDavStoragePlugin.kt:96)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at com.stevesoltys.seedvault.transport.backup.BackupCoordinator.initializeDevice(BackupCoordinator.kt:128)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at com.stevesoltys.seedvault.transport.ConfigurableBackupTransport$initializeDevice$1.invokeSuspend(ConfigurableBackupTransport.kt:128)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:280)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at com.stevesoltys.seedvault.transport.ConfigurableBackupTransport.initializeDevice(ConfigurableBackupTransport.kt:127)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at android.app.backup.BackupTransport$TransportImpl.initializeDevice(BackupTransport.java:758)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at com.android.internal.backup.IBackupTransport$Stub.onTransact(IBackupTransport.java:649)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at android.os.Binder.execTransactInternal(Binder.java:1501)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at android.os.Binder.execTransact(Binder.java:1440)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: Caused by: at.bitfire.dav4jvm.exception.DavException: Received redirect from HTTPS to HTTP
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at at.bitfire.dav4jvm.DavResource.followRedirects$build(DavResource.kt:616)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at at.bitfire.dav4jvm.DavResource.head(DavResource.kt:272)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	at com.stevesoltys.seedvault.plugins.webdav.WebDavStoragePlugin.initializeDevice(WebDavStoragePlugin.kt:88)
06-02 11:42:40.645 30094 30665 E BackupCoordinator: 	... 15 more
06-02 11:42:40.649  2053  4288 E BackupManagerService: Transport error in initializeDevice()
06-02 11:42:40.650 30094 30665 I BackupCoordinator: Request incremental backup time. Returned 0
06-02 11:42:40.651  2053  4288 W BackupManagerService: Init failed on com.stevesoltys.seedvault.transport.ConfigurableBackupTransport resched in 0
06-02 11:42:40.653  2053  2053 V BackupManagerService: Running a device init; 1 pending
06-02 11:42:40.653  2053  2053 V BackupManagerService: [UserID:0] initializeTransport(): [com.stevesoltys.seedvault.transport.ConfigurableBackupTransport]
06-02 11:42:40.653  2053  2053 V BackupManagerService: [UserID:0] Acquired wakelock:*backup*-0-4288
06-02 11:42:40.653  2053  4288 V BackupManagerService: [UserID:0] Released wakelock:*backup*-0-4288
06-02 11:42:40.654  2053  4288 I BackupManagerService: Initializing (wiping) backup transport storage: com.stevesoltys.seedvault.transport.ConfigurableBackupTransport
06-02 11:42:40.655 30094 30094 I BackupStateManager: appBackupRunning: false, filesBackupRunning: false, workInfoState: CANCELLED
06-02 11:42:40.656 30094 30094 D ConfigurableBackupTransportService: Service destroyed.
06-02 11:42:40.661 30094 30094 D ConfigurableBackupTransportService: Service created.
06-02 11:42:40.663 30094 30094 D ConfigurableBackupTransportService: Transport bound.

```

This is a valid webdav endpoint, works well with Finder network mount on MacOS and other places.

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/StoragePluginManager.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/StoragePluginManager.kt
@@ -139,7 +139,7 @@ class StoragePluginManager(
     suspend fun getFreeSpace(): Long? {
         return try {
             appPlugin.getFreeSpace()
-        } catch (e: Exception) {
+        } catch (e: Throwable) { // NoClassDefFound isn't an [Exception], can get thrown by dav4jvm
             Log.e("StoragePluginManager", "Error getting free space: ", e)
             null
         }

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStorage.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStorage.kt
@@ -142,7 +142,7 @@ internal abstract class WebDavStorage(
                 callback = callback,
             )
         } catch (e: HttpException) {
-            if (e.code == 400) {
+            if (e.isUnsupportedPropfind()) {
                 Log.i(TAG, "Got ${e.response}, trying two depth=1 PROPFINDs...")
                 propfindFakeTwo(callback)
             } else {
@@ -167,6 +167,18 @@ internal abstract class WebDavStorage(
                 )
             }
         }
+    }
+
+    protected fun HttpException.isUnsupportedPropfind(): Boolean {
+        // nginx returns 400 for depth=2
+        if (code == 400) {
+            return true
+        }
+        // lighttpd returns 403 with <DAV:propfind-finite-depth/> error as if we used infinity
+        if (code == 403 && responseBody?.contains("propfind-finite-depth") == true) {
+            return true
+        }
+        return false
     }
 
     protected suspend fun DavCollection.createFolder(xmlBody: String? = null): okhttp3.Response {

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStorage.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStorage.kt
@@ -9,8 +9,12 @@ import android.util.Log
 import at.bitfire.dav4jvm.BasicDigestAuthHandler
 import at.bitfire.dav4jvm.DavCollection
 import at.bitfire.dav4jvm.MultiResponseCallback
+import at.bitfire.dav4jvm.Property
+import at.bitfire.dav4jvm.PropertyFactory
+import at.bitfire.dav4jvm.PropertyRegistry
 import at.bitfire.dav4jvm.Response
 import at.bitfire.dav4jvm.Response.HrefRelation.SELF
+import at.bitfire.dav4jvm.XmlUtils.NS_WEBDAV
 import at.bitfire.dav4jvm.exception.HttpException
 import at.bitfire.dav4jvm.property.DisplayName
 import at.bitfire.dav4jvm.property.ResourceType
@@ -27,6 +31,7 @@ import okhttp3.OkHttpClient
 import okhttp3.RequestBody
 import okhttp3.internal.closeQuietly
 import okio.BufferedSink
+import org.xmlpull.v1.XmlPullParser
 import java.io.IOException
 import java.io.InputStream
 import java.io.OutputStream
@@ -68,6 +73,10 @@ internal abstract class WebDavStorage(
 
     protected val baseUrl = webDavConfig.url
     protected val url = "${webDavConfig.url}/$root"
+
+    init {
+        PropertyRegistry.register(GetLastModified.Factory)
+    }
 
     @Throws(IOException::class)
     protected suspend fun getOutputStream(location: HttpUrl): OutputStream {
@@ -236,4 +245,20 @@ internal abstract class WebDavStorage(
         }
     }
 
+}
+
+/**
+ * A fake version of [at.bitfire.dav4jvm.property.GetLastModified] which we register
+ * so we don't need to depend on `org.apache.commons.lang3` which is used for date parsing.
+ */
+class GetLastModified : Property {
+    companion object {
+        @JvmField
+        val NAME = Property.Name(NS_WEBDAV, "getlastmodified")
+    }
+
+    object Factory : PropertyFactory {
+        override fun getName() = NAME
+        override fun create(parser: XmlPullParser): GetLastModified? = null
+    }
 }

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStorage.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStorage.kt
@@ -8,7 +8,11 @@ package com.stevesoltys.seedvault.plugins.webdav
 import android.util.Log
 import at.bitfire.dav4jvm.BasicDigestAuthHandler
 import at.bitfire.dav4jvm.DavCollection
+import at.bitfire.dav4jvm.MultiResponseCallback
 import at.bitfire.dav4jvm.Response
+import at.bitfire.dav4jvm.Response.HrefRelation.SELF
+import at.bitfire.dav4jvm.exception.HttpException
+import at.bitfire.dav4jvm.property.DisplayName
 import at.bitfire.dav4jvm.property.ResourceType
 import kotlinx.coroutines.DelicateCoroutinesApi
 import kotlinx.coroutines.Dispatchers
@@ -122,6 +126,47 @@ internal abstract class WebDavStorage(
             }
         }
         return pipedInputStream
+    }
+
+    /**
+     * Tries to do [DavCollection.propfind] with a depth of `2` which is not in RFC4918.
+     * Since `infinity` isn't supported by nginx either,
+     * we fallback to iterating over all folders found with depth `1`
+     * and do another PROPFIND on those, passing the given [callback].
+     */
+    protected fun DavCollection.propfindDepthTwo(callback: MultiResponseCallback) {
+        try {
+            propfind(
+                depth = 2, // this isn't defined in RFC4918
+                reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
+                callback = callback,
+            )
+        } catch (e: HttpException) {
+            if (e.code == 400) {
+                Log.i(TAG, "Got ${e.response}, trying two depth=1 PROPFINDs...")
+                propfindFakeTwo(callback)
+            } else {
+                throw e
+            }
+        }
+    }
+
+    private fun DavCollection.propfindFakeTwo(callback: MultiResponseCallback) {
+        propfind(
+            depth = 1,
+            reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
+        ) { response, relation ->
+            debugLog { "propFindFakeTwo() = $response" }
+            // This callback will be called for everything in the folder
+            callback.onResponse(response, relation)
+            if (relation != SELF && response.isFolder()) {
+                DavCollection(okHttpClient, response.href).propfind(
+                    depth = 1,
+                    reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
+                    callback = callback,
+                )
+            }
+        }
     }
 
     protected suspend fun DavCollection.createFolder(xmlBody: String? = null): okhttp3.Response {

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
@@ -163,7 +163,7 @@ internal class WebDavStoragePlugin(
     override suspend fun getAvailableBackups(): Sequence<EncryptedMetadata>? {
         return try {
             doGetAvailableBackups()
-        } catch (e: Exception) {
+        } catch (e: Throwable) { // NoClassDefFound isn't an [Exception], can get thrown by dav4jvm
             Log.e(TAG, "Error getting available backups: ", e)
             null
         }

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
@@ -9,6 +9,7 @@ import android.content.Context
 import android.util.Log
 import at.bitfire.dav4jvm.DavCollection
 import at.bitfire.dav4jvm.Response.HrefRelation.SELF
+import at.bitfire.dav4jvm.exception.HttpException
 import at.bitfire.dav4jvm.exception.NotFoundException
 import at.bitfire.dav4jvm.property.DisplayName
 import at.bitfire.dav4jvm.property.QuotaAvailableBytes
@@ -174,20 +175,25 @@ internal class WebDavStoragePlugin(
 
         // get all restore set tokens in root folder
         val tokens = ArrayList<Long>()
-        davCollection.propfind(
-            depth = 2,
-            reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
-        ) { response, relation ->
-            debugLog { "getAvailableBackups() = $response" }
-            // This callback will be called for every file in the folder
-            if (relation != SELF && !response.isFolder() && response.href.pathSize >= 2 &&
-                response.hrefName() == FILE_BACKUP_METADATA
-            ) {
-                val tokenName = response.href.pathSegments[response.href.pathSegments.size - 2]
-                getTokenOrNull(tokenName)?.let { token ->
-                    tokens.add(token)
+        try {
+            davCollection.propfind(
+                depth = 2,
+                reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
+            ) { response, relation ->
+                debugLog { "getAvailableBackups() = $response" }
+                // This callback will be called for every file in the folder
+                if (relation != SELF && !response.isFolder() && response.href.pathSize >= 2 &&
+                    response.hrefName() == FILE_BACKUP_METADATA
+                ) {
+                    val tokenName = response.href.pathSegments[response.href.pathSegments.size - 2]
+                    getTokenOrNull(tokenName)?.let { token ->
+                        tokens.add(token)
+                    }
                 }
             }
+        } catch (e: HttpException) {
+            if (e.code == 400) getBackupTokenWithDepthOne(davCollection, tokens)
+            else throw e
         }
         val tokenIterator = tokens.iterator()
         return generateSequence {
@@ -195,6 +201,34 @@ internal class WebDavStoragePlugin(
             val token = tokenIterator.next()
             EncryptedMetadata(token) {
                 getInputStream(token, FILE_BACKUP_METADATA)
+            }
+        }
+    }
+
+    private fun getBackupTokenWithDepthOne(davCollection: DavCollection, tokens: ArrayList<Long>) {
+        davCollection.propfind(
+            depth = 1,
+            reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
+        ) { response, relation ->
+            debugLog { "getBackupTokenWithDepthOne() = $response" }
+
+            // we are only interested in sub-folders, skip rest
+            if (relation == SELF || !response.isFolder()) return@propfind
+
+            val token = getTokenOrNull(response.hrefName()) ?: return@propfind
+            val tokenUrl = response.href.newBuilder()
+                .addPathSegment(FILE_BACKUP_METADATA)
+                .build()
+            // check if .backup.metadata file exists using HEAD request,
+            // because some servers (e.g. nginx don't list hidden files with PROPFIND)
+            try {
+                DavCollection(okHttpClient, tokenUrl).head {
+                    debugLog { "getBackupTokenWithDepthOne() = $response" }
+                    tokens.add(token)
+                }
+            } catch (e: Exception) {
+                // just log exception and continue, we want to find all files that are there
+                Log.e(TAG, "Error retrieving $tokenUrl: ", e)
             }
         }
     }

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
@@ -34,7 +34,7 @@ internal class WebDavStoragePlugin(
 ) : WebDavStorage(webDavConfig, root), StoragePlugin<WebDavConfig> {
 
     override suspend fun test(): Boolean {
-        val location = baseUrl.toHttpUrl()
+        val location = (if (baseUrl.endsWith('/')) baseUrl else "$baseUrl/").toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
 
         val webDavSupported = suspendCoroutine { cont ->
@@ -50,7 +50,7 @@ internal class WebDavStoragePlugin(
     }
 
     override suspend fun getFreeSpace(): Long? {
-        val location = url.toHttpUrl()
+        val location = "$url/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
 
         val availableBytes = suspendCoroutine { cont ->
@@ -70,7 +70,7 @@ internal class WebDavStoragePlugin(
 
     @Throws(IOException::class)
     override suspend fun startNewRestoreSet(token: Long) {
-        val location = "$url/$token".toHttpUrl()
+        val location = "$url/$token/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
 
         val response = davCollection.createFolder()
@@ -81,7 +81,7 @@ internal class WebDavStoragePlugin(
     override suspend fun initializeDevice() {
         // TODO does it make sense to delete anything
         //  when [startNewRestoreSet] is always called first? Maybe unify both calls?
-        val location = url.toHttpUrl()
+        val location = "$url/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
 
         try {
@@ -169,7 +169,7 @@ internal class WebDavStoragePlugin(
     }
 
     private suspend fun doGetAvailableBackups(): Sequence<EncryptedMetadata> {
-        val location = url.toHttpUrl()
+        val location = "$url/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
 
         // get all restore set tokens in root folder

--- a/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePlugin.kt
@@ -192,7 +192,7 @@ internal class WebDavStoragePlugin(
                 }
             }
         } catch (e: HttpException) {
-            if (e.code == 400) getBackupTokenWithDepthOne(davCollection, tokens)
+            if (e.isUnsupportedPropfind()) getBackupTokenWithDepthOne(davCollection, tokens)
             else throw e
         }
         val tokenIterator = tokens.iterator()

--- a/app/src/main/java/com/stevesoltys/seedvault/storage/WebDavStoragePlugin.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/storage/WebDavStoragePlugin.kt
@@ -77,10 +77,7 @@ internal class WebDavStoragePlugin(
         val chunkIds = ArrayList<String>()
         try {
             val duration = measureDuration {
-                davCollection.propfind(
-                    depth = 2,
-                    reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
-                ) { response, relation ->
+                davCollection.propfindDepthTwo { response, relation ->
                     debugLog { "getAvailableChunkIds() = $response" }
                     // This callback will be called for every file in the folder
                     if (relation != SELF && response.isFolder()) {
@@ -162,13 +159,10 @@ internal class WebDavStoragePlugin(
 
         val snapshots = ArrayList<StoredSnapshot>()
         try {
-            davCollection.propfind(
-                depth = 2,
-                reqProp = arrayOf(DisplayName.NAME, ResourceType.NAME),
-            ) { response, relation ->
+            davCollection.propfindDepthTwo { response, relation ->
                 debugLog { "getBackupSnapshotsForRestore() = $response" }
                 // This callback will be called for every file in the folder
-                if (relation != SELF && !response.isFolder()) {
+                if (relation != SELF && !response.isFolder() && response.href.pathSize >= 2) {
                     val name = response.hrefName()
                     val match = snapshotRegex.matchEntire(name)
                     if (match != null) {

--- a/app/src/main/java/com/stevesoltys/seedvault/storage/WebDavStoragePlugin.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/storage/WebDavStoragePlugin.kt
@@ -49,7 +49,7 @@ internal class WebDavStoragePlugin(
 
     @Throws(IOException::class)
     override suspend fun init() {
-        val location = url.toHttpUrl()
+        val location = "$url/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
 
         try {
@@ -67,7 +67,7 @@ internal class WebDavStoragePlugin(
 
     @Throws(IOException::class)
     override suspend fun getAvailableChunkIds(): List<String> {
-        val location = "$url/$folder".toHttpUrl()
+        val location = "$url/$folder/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
         debugLog { "getAvailableChunkIds($location)" }
 
@@ -117,7 +117,7 @@ internal class WebDavStoragePlugin(
     ) {
         val s = missingChunkFolders.size
         for ((i, chunkFolderName) in missingChunkFolders.withIndex()) {
-            val location = "$url/$folder/$chunkFolderName".toHttpUrl()
+            val location = "$url/$folder/$chunkFolderName/".toHttpUrl()
             val davCollection = DavCollection(okHttpClient, location)
             val response = davCollection.createFolder()
             debugLog { "Created missing folder $chunkFolderName (${i + 1}/$s) $response" }
@@ -156,7 +156,7 @@ internal class WebDavStoragePlugin(
 
     @Throws(IOException::class)
     override suspend fun getBackupSnapshotsForRestore(): List<StoredSnapshot> {
-        val location = url.toHttpUrl()
+        val location = "$url/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
         debugLog { "getBackupSnapshotsForRestore($location)" }
 
@@ -220,7 +220,7 @@ internal class WebDavStoragePlugin(
 
     @Throws(IOException::class)
     override suspend fun getCurrentBackupSnapshots(): List<StoredSnapshot> {
-        val location = "$url/$folder".toHttpUrl()
+        val location = "$url/$folder/".toHttpUrl()
         val davCollection = DavCollection(okHttpClient, location)
         debugLog { "getCurrentBackupSnapshots($location)" }
 

--- a/app/src/test/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePluginTest.kt
+++ b/app/src/test/java/com/stevesoltys/seedvault/plugins/webdav/WebDavStoragePluginTest.kt
@@ -71,6 +71,10 @@ internal class WebDavStoragePluginTest : TransportTest() {
         plugin.getOutputStream(token, FILE_BACKUP_METADATA).use {
             it.write(metadata)
         }
+
+        // now we have data
+        assertTrue(plugin.hasData(token, FILE_BACKUP_METADATA))
+
         try {
             // now we have one backup matching our token
             val backups = plugin.getAvailableBackups()?.toSet() ?: fail()

